File: packages/litertlm-native/ios/SessionSnapshotStore.swift
import Foundation

/// Phase 2 S4 — KV snapshot metadata (native serialize unavailable → message replay fallback).
enum SessionSnapshotStore {
  struct SnapshotMeta {
    let conversationId: String
    let usedNativeKvSerialize: Bool
    let messageCount: Int
    let updatedAt: TimeInterval
  }

  private static let metaSuffix = ".kvsnapshot.meta.json"
  private static let snapshotSuffix = ".kvsnapshot"

  static func inferenceDir(cacheDir: String) -> URL {
    URL(fileURLWithPath: cacheDir).appendingPathComponent("inference", isDirectory: true)
  }

  static func metaFile(cacheDir: String, conversationId: String) -> URL {
    inferenceDir(cacheDir: cacheDir).appendingPathComponent("\(conversationId)\(metaSuffix)")
  }

  static func snapshotFile(cacheDir: String, conversationId: String) -> URL {
    inferenceDir(cacheDir: cacheDir).appendingPathComponent("\(conversationId)\(snapshotSuffix)")
  }

  static func writeMeta(
    cacheDir: String,
    conversationId: String,
    messageCount: Int,
    usedNativeKvSerialize: Bool = false
  ) throws -> SnapshotMeta {
    let dir = inferenceDir(cacheDir: cacheDir)
    try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    let meta = SnapshotMeta(
      conversationId: conversationId,
      usedNativeKvSerialize: usedNativeKvSerialize,
      messageCount: messageCount,
      updatedAt: Date().timeIntervalSince1970 * 1000
    )
    let payload: [String: Any] = [
      "version": 1,
      "conversationId": meta.conversationId,
      "usedNativeKvSerialize": meta.usedNativeKvSerialize,
      "messageCount": meta.messageCount,
      "updatedAt": Int(meta.updatedAt),
    ]
    let data = try JSONSerialization.data(withJSONObject: payload)
    try data.write(to: metaFile(cacheDir: cacheDir, conversationId: conversationId))
    try Data().write(to: snapshotFile(cacheDir: cacheDir, conversationId: conversationId))
    return meta
  }

  static func readMeta(cacheDir: String, conversationId: String) -> SnapshotMeta? {
    let url = metaFile(cacheDir: cacheDir, conversationId: conversationId)
    guard let data = try? Data(contentsOf: url),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let id = json["conversationId"] as? String
    else {
      return nil
    }
    return SnapshotMeta(
      conversationId: id,
      usedNativeKvSerialize: json["usedNativeKvSerialize"] as? Bool ?? false,
      messageCount: json["messageCount"] as? Int ?? 0,
      updatedAt: TimeInterval(json["updatedAt"] as? Int ?? 0)
    )
  }

  static func hasSnapshot(cacheDir: String, conversationId: String) -> Bool {
    FileManager.default.fileExists(atPath: metaFile(cacheDir: cacheDir, conversationId: conversationId).path)
  }

  static func deleteSnapshot(cacheDir: String, conversationId: String) {
    try? FileManager.default.removeItem(at: metaFile(cacheDir: cacheDir, conversationId: conversationId))
    try? FileManager.default.removeItem(at: snapshotFile(cacheDir: cacheDir, conversationId: conversationId))
  }

  static func restoredFrom(_ meta: SnapshotMeta?) -> String {
    guard let meta else {
      return "empty"
    }
    if meta.usedNativeKvSerialize {
      return "kv_snapshot"
    }
    if meta.messageCount > 0 {
      return "message_replay"
    }
    return "empty"
  }
}

File: packages/litertlm-native/android/src/main/java/expo/modules/litertlmnative/SessionSnapshotStore.kt
package expo.modules.litertlmnative

import org.json.JSONObject
import java.io.File

/**
 * Phase 2 S4 — KV snapshot metadata (native serialize unavailable → message replay fallback).
 * Binary `.kvsnapshot` is reserved for future LiteRT-LM KV serialize; metadata drives JS replay.
 */
object SessionSnapshotStore {
  private const val META_SUFFIX = ".kvsnapshot.meta.json"
  private const val SNAPSHOT_SUFFIX = ".kvsnapshot"

  data class SnapshotMeta(
    val conversationId: String,
    val usedNativeKvSerialize: Boolean,
    val messageCount: Int,
    val updatedAt: Long,
  )

  fun inferenceDir(cacheDir: String): File = File(cacheDir, "inference")

  fun metaFile(cacheDir: String, conversationId: String): File =
    File(inferenceDir(cacheDir), "$conversationId$META_SUFFIX")

  fun snapshotFile(cacheDir: String, conversationId: String): File =
    File(inferenceDir(cacheDir), "$conversationId$SNAPSHOT_SUFFIX")

  fun writeMeta(
    cacheDir: String,
    conversationId: String,
    messageCount: Int,
    usedNativeKvSerialize: Boolean = false,
  ): SnapshotMeta {
    val dir = inferenceDir(cacheDir)
    dir.mkdirs()
    val meta =
      SnapshotMeta(
        conversationId = conversationId,
        usedNativeKvSerialize = usedNativeKvSerialize,
        messageCount = messageCount,
        updatedAt = System.currentTimeMillis(),
      )
    val json =
      JSONObject()
        .put("version", 1)
        .put("conversationId", meta.conversationId)
        .put("usedNativeKvSerialize", meta.usedNativeKvSerialize)
        .put("messageCount", meta.messageCount)
        .put("updatedAt", meta.updatedAt)
    metaFile(cacheDir, conversationId).writeText(json.toString())
    // Placeholder path for future native KV bytes.
    snapshotFile(cacheDir, conversationId).writeText("")
    return meta
  }

  fun readMeta(cacheDir: String, conversationId: String): SnapshotMeta? {
    val file = metaFile(cacheDir, conversationId)
    if (!file.exists()) {
      return null
    }
    return runCatching {
      val json = JSONObject(file.readText())
      SnapshotMeta(
        conversationId = json.getString("conversationId"),
        usedNativeKvSerialize = json.optBoolean("usedNativeKvSerialize", false),
        messageCount = json.optInt("messageCount", 0),
        updatedAt = json.optLong("updatedAt", 0L),
      )
    }.getOrNull()
  }

  fun hasSnapshot(cacheDir: String, conversationId: String): Boolean =
    metaFile(cacheDir, conversationId).exists()

  fun deleteSnapshot(cacheDir: String, conversationId: String) {
    metaFile(cacheDir, conversationId).delete()
    snapshotFile(cacheDir, conversationId).delete()
  }

  fun restoredFrom(meta: SnapshotMeta?): String =
    when {
      meta == null -> "empty"
      meta.usedNativeKvSerialize -> "kv_snapshot"
      meta.messageCount > 0 -> "message_replay"
      else -> "empty"
    }
}
